### PR TITLE
[Feature/pur-324-upselltrack]: PDP Upsell functions

### PR DESCRIPTION
**Pull request by kevin-ho87**
Ticket
https://aesoponline.atlassian.net/browse/PUR-324

PDP Upsell expose functions for tracking when:
- scroll into view
- onClick

### CODE REVIEW

File: src/components/ProductDetailHeader/ProductDetailHeader.tsx [hunk 1]
```diff
@@ -33,6 +35,8 @@ const ProductDetailHeader: ProductDetailHeaderType = ({
     onBreadcrumbClick(item, selectedVariant);
   const handleOnFlyinOpenCloseClick = (flyinState) =>
     onFlyinOpenCloseClick(flyinState, selectedVariant);
+  const handleOnUpsellScrollIntoView = () =>
+    onUpsellScrollIntoView && onUpsellScrollIntoView();
```
Comment (MkMan) [inline src/components/ProductDetailHeader/ProductDetailHeader.tsx:39]: Would this function return `false` if `handleOnUpsellScrollIntoView` isn't provided?

File: src/components/ProductDetailHeader/components/ProductDetailBody/ProductDetailBody.tsx [hunk 2]
```diff
@@ -27,11 +28,25 @@ const ProductDetailBody: ProductDetailBodyType = ({
...
+    if (onUpsellScrollIntoView) {
+      onUpsellScrollIntoView();
+    }
+  };
+
+  useExecuteOnImpression(upsellRef, handleOnUpsellScrollIntoView, {
+    threshold: 0.2,
+    isExecutableOnReEntry: false,
+  });
+
```
Comment (zack76) [inline src/components/ProductDetailHeader/components/ProductDetailBody/ProductDetailBody.tsx:49]: how about 
```
if (onUpsellScrollIntoView) {
     useExecuteOnImpression(upsellRef, onUpsellScrollIntoView, {
     threshold: 0.2,
     isExecutableOnReEntry: false,
    });
}
```

**Review by zack76 — APPROVED** [review]

File: src/components/ProductDetailHeader/components/ProductDetailBody/ProductDetailBody.tsx (hunk 2 above)
Comment (MkMan) [reply src/components/ProductDetailHeader/components/ProductDetailBody/ProductDetailBody.tsx:49]: Would that violate the [rule](https://reactjs.org/docs/hooks-rules.html#only-call-hooks-at-the-top-level) on calling a hook inside a conditional?

Comment (zack76) [reply src/components/ProductDetailHeader/components/ProductDetailBody/ProductDetailBody.tsx:49]: ah sorry, i didnt realise that

File: src/components/ProductDetailHeader/ProductDetailHeader.tsx (hunk 1 above)
Comment (kevin-ho87) [reply src/components/ProductDetailHeader/ProductDetailHeader.tsx:39]: I think it just returns `undefined` if `onUpsellScrollIntoView` prop is not provided

Comment (kevin-ho87) [reply src/components/ProductDetailHeader/ProductDetailHeader.tsx:39]:
```
const handleOnUpsellScrollIntoView = () =>
    !!onUpsellScrollIntoView && onUpsellScrollIntoView();
```
Would adding `!!` work better?

Comment (MkMan) [reply src/components/ProductDetailHeader/ProductDetailHeader.tsx:39]: I think what you already have is better 👍 `!!` would make `onUpsellScrollIntoView` into `false` if it was not provided and the function will then return `false`, right?

**Pull request merged**